homeassistant/components/elkm1/__init__.py

```diff
@@ -0,0 +1,220 @@
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+SUPPORTED_DOMAINS = ['alarm_control_panel']
+
+
+async def async_setup(hass: HomeAssistant, hass_config: ConfigType) -> bool:
+    """Set up the Elk M1 platform."""
+
+    from elkm1_lib.const import Max
+    from elkm1_lib.message import housecode_to_index
+    import elkm1_lib as elkm1
+
+    configs = {
+        CONF_AREA: Max.AREAS.value,
+        CONF_COUNTER: Max.COUNTERS.value,
+        CONF_KEYPAD: Max.KEYPADS.value,
+        CONF_OUTPUT: Max.OUTPUTS.value,
+        CONF_PANEL: 1,
+        CONF_PLC: Max.LIGHTS.value,
+        CONF_SETTING: Max.SETTINGS.value,
+        CONF_TASK: Max.TASKS.value,
+        CONF_THERMOSTAT: Max.THERMOSTATS.value,
+        CONF_USER: Max.USERS.value,
+        CONF_ZONE: Max.ZONES.value,
+    }
+
+    def parse_value(val, max_):
+        """Parse a value as an int or housecode."""
+        i = int(val) if val.isdigit() else (housecode_to_index(val) + 1)
+        if i < 1 or i > max_:
+            raise vol.Invalid('Value not in range 1-%d: "%s"' % (max_, val))
+        return i
+
+    def parse_range(config, item, set_to, values, max_):
+        """Parse a range list, e.g. range in form of 3 or 2-7"""
+        ranges = config.get(item, [])
+        for rng in ranges:
+            rng = str(rng)
+            if '-' in rng:
+                rng_vals = [s.strip() for s in rng.split('-')]
+                start = parse_value(rng_vals[0], max_)
+                end = parse_value(rng_vals[1], max_)
+            else:
+                start = end = parse_value(rng, max_)
+            values[start-1:end] = [set_to] * (end - start + 1)
+
+    def parse_config(item, max_):
+        """Parse a config for an element type such as: zones, plc, etc."""
+        if item not in config_raw:
+            return (True, [True] * max_)
+
+        conf = config_raw[item]
+
+        if CONF_ENABLED in conf and not conf[CONF_ENABLED]:
+            return (False, [False] * max_)
+
+        included = [CONF_INCLUDE not in conf] * max_
+        parse_range(conf, CONF_INCLUDE, True, included, max_)
+        parse_range(conf, CONF_EXCLUDE, False, included, max_)
+
+        return (True, included)
+
+    config_raw = hass_config.get(DOMAIN)
+    config = {}
+
+    host = config_raw[CONF_HOST]
+    username = config_raw.get(CONF_USERNAME)
+    password = config_raw.get(CONF_PASSWORD)
+    if host.startswith('elks:') and (username is None or password is None):
+        raise vol.Invalid("Specify username & password for elks://")
+
+    for item, max_ in configs.items():
+        config[item] = {}
+        (config[item]['enabled'], config[item]['included']) = \
+            parse_config(item, max_)
+
+    config['temperature_unit'] = config_raw[CONF_TEMPERATURE_UNIT]
+
+    elk = elkm1.Elk({'url': host, 'userid': username, 'password': password})
+    elk.connect()
+
+    hass.data[DOMAIN] = {'elk': elk, 'config': config,
+                         'entities': {}, 'keypads': {}}
```

Comment:
Are we still using `'entities'` and `'keypads'` keys?

Reply:
entities no.

keypads will be used as soon as the keypad sensor is submitted as a PR